### PR TITLE
Events v2

### LINKED ISSUE #82: Agenda branding: End hour not shown for multi-date events
In the events detail screen

### LINKED ISSUE #436: Support multi-day events for uitdatabank syncing
See https://github.com/our-city-app/oca-backend/blob/5ecbcd9d343f9ae9ea558ae8bb4b2bb4b910100f/src/solutions/common/bizz/events.py#L271

### LINKED ISSUE #1440: UiT improvement for periods
There are events without start time, but with a period. Let's find a solution for them when reworking the UiT agenda.

```
...
INFO     2019-09-04 08:03:01,636 uitdatabank.py:401]
Skipping event because it had no starttime (list) for:

...
 u'calendar': {u'periods': {u'period': {u'datefrom': u'2019-09-17',
                                        u'dateto': u'2019-10-15'}}},
...
```

### LINKED ISSUE #1460: New events UI & backend logic
- 1 embedded app shared over every service in every app
    - Use app primary & secondary colors to theme the UI https://github.com/ionic-team/ionic-docs/blob/master/src/components/color-gen/color.ts
- add embedded app to service menu item (no version)
- events their dates need to be refactored so they're basically the same as how uit returns them. - Sort events by start date?

### LINKED ISSUE #1479: Embedded app with events and news
- [x] #1440 Improve data structure so we can handle periodic events better
    - [x] Change dates structure
    - ~[ ] Save 'event type' (cursus/creativiteit/gezondheid/voeding/...) or whatever it's called so we can filter on it. No event types for google calendar events~ later 
- [x] Make it possible to open embedded app from service menu item. Embedded app should receive some context (the menu item that was used to open it)
    - [x] Server + web UI
    - [x] Android
    - [x] iOS
- [x] Embedded app should be able to execute getNewsGroups and getNewsItems calls (bridge between javascript and ios/android rpc calls, use some generic handler that can be easily reused like the IntentResponseHandler on android)
   - [x] Android
   - [x] iOS
- [x] Method to open a specific news group from js (extend rogerthat.open)
- [x] #1479 Create embedded app to be used for various OCA stuff (starting with events). No actual functionality yet.
- [x] List of agenda events
- [x] List of news items
- [x] Filter agenda events page 
  - textfield
  - time: Vandaag, Morgen, dit weekend, Volgende 7 dagen, Volgende 14 dagen, Volgende 30 dagen, Kies een datum
- [x] Index events in search index (id, title, description, every timeframe)
- [x] Event detail page

Screenshot 2019-10-01 at 14 16 30

### LINKED ISSUE #1514: events: search
Instead of using datastore to search for events, use elasticsearch.
Search will happen with these params:
  - event date > current date
  - one or more filters, see https://documentatie.uitdatabank.be/content/json-ld-crud-api/latest/events/event-terms.html
Index the event in such a way that you don't need to update the index every day

Not sure how deep we want to go with the filters.

### LINKED ISSUE #1521: events: cleanup & migration
- migrate all services so that the 'agenda' menu item is the 'oca' embedded app instead of a branding ( = republish all services with 'agenda' module)
- remove EventGuest and EventReminder models
- remove / hard delete all brandings that aren't used with name 'Events App'

### PR DESCRIPTION
**Pull request by lucasvanhalst**
Closes #82
Closes #436
Closes #1440
Closes #1460
Closes #1479
Closes #1514
Closes #1521

Screenshot 2019-12-18 at 16 50 23
Screenshot 2019-12-18 at 16 50 29

To test, install the updated version of the cordova plugin in the embedded-apps/oca folder  `npm install git+https://github.com/our-city-app/cordova-mobicage-plugin.git#news_apis`

### CODE REVIEW

**Comment by lucasvanhalst** [comment]
remarks:

- [ ] reduce thumbnail size
- [ ] markdown description in app
- [ ] location -> convert to item with icon
- [ ] parse css in description of uit events https://www.uitinvlaanderen.be/agenda/e/19-kerstwandeling/d6953545-e98c-4ca7-9a03-1d674e8a0062

**Comment by lucasvanhalst** [comment]
TODO: show message to update app in case app version < supported version

**Comment by RubenMattan** [comment]
- make compatible with current android/ios clients, so hide content (news) that is not available
- Allow saving an event that starts on 20:00 en ends on 01:00 via the dashboard

File: embedded-apps/oca/src/app/app.component.ts [hunk 1]
```diff
@@ -35,27 +37,37 @@ export class AppComponent {
...
             this.statusBar.backgroundColorByHexString(rogerthat.system.colors.primaryDark);
           }
         } else {
           this.statusBar.styleDefault();
         }
         this.rogerthatService.initialize();
-        this.rogerthatService.getContext().subscribe(context => {
-          this.router.navigate(this.getRootPage(context));
-        });
+        if (!rogerthat.system.debug && !this.rogerthatService.isSupported([2, 1, 5677], [2, 1, 3978])) {
```
Comment (RubenMattan) [inline embedded-apps/oca/src/app/app.component.ts]: Update version when released

File: embedded-apps/oca/src/global.scss [hunk 2]
```diff
@@ -17,10 +17,10 @@
 @import '~@ionic/angular/css/display.css';
 /* Optional CSS utils that can be commented out */
 @import "~@ionic/angular/css/padding.css";
-@import "~@ionic/angular/css/float-elements.css";
-@import "~@ionic/angular/css/text-alignment.css";
-@import "~@ionic/angular/css/text-transformation.css";
-@import "~@ionic/angular/css/flex-utils.css";
+//@import "~@ionic/angular/css/float-elements.css";
```
Comment (RubenMattan) [inline embedded-apps/oca/src/global.scss]: debug?

File: src/solutions/common/cron/events/events.py [hunk 3]
```diff
@@ -19,119 +19,31 @@
...
-            reminder.status = EventReminder.STATUS_REMINDED
-            reminder.put()
-    else:
-        reminder.status = EventReminder.STATUS_REMINDED
-        reminder.put()
+        yesterday = datetime.datetime.now() - relativedelta.relativedelta(days=1)
+        event_keys = Event.list_less_than_end_date(yesterday).fetch(keys_only=True)
+        logging.info('Cleaning up %d expired Event models' % len(event_keys))
+        ndb.delete_multi(event_keys)
+        delete_events_from_index(event_keys)
```
Comment (RubenMattan) [inline src/solutions/common/cron/events/events.py]: move before ndb.delete

File: src/solutions/common/bizz/events/__init__.py [hunk 4]
```diff
@@ -0,0 +1,667 @@
...
+    settings = get_solution_settings(service_user)
+    event = get_event_by_id(service_user, settings.solution, event_id)
+    if event:
+        if event.source == Event.SOURCE_CMS:
+            remove_files([media.ref for media in event.media if media.ref])
+            event.key.delete()
+        else:
+            event.deleted = True
+            event.put()
+        delete_events_from_index([event.key])
```
Comment (RubenMattan) [inline src/solutions/common/bizz/events/__init__.py]: move before event delete

File: src/solutions/common/cron/associations.py [hunk 5]
```diff
@@ -58,8 +59,9 @@ def update_statistic():
 
     future_event_count_dict = {}
 
-    for future_event_key in Event.get_future_event_keys(current_date):
-        service_email = future_event_key.parent().name()
+    event_keys = Event.list_greater_than_start_date(current_date)
```
Comment (RubenMattan) [inline src/solutions/common/cron/associations.py]: missing deleted = False

**Comment by lucasvanhalst** [comment]
- [x] migration start 21:00 end 05:00 results in end date before start date instead of next day
- make compatible with current android/ios clients, so hide content (news) that is not available
- Allow saving an event that starts on 20:00 en ends on 01:00 via the dashboard
    - that doesn't make any sense?
- improve search (you have 4 events [event 1, event 2, event 3, event 4] when searching for event 4 you shoudn't see the other ones)
    - that would actually make it worse, not improve it?
- nuke SolutionAutoBroadcastTypes for module agenda
- nuke SolutionCalendarAdmin
- (optional) circle close button remains visible for a few seconds after click and returning to list
- [x] resolve conflicts

File: embedded-apps/oca/src/global.scss (hunk 2 above)
Comment (lucasvanhalst) [reply embedded-apps/oca/src/global.scss]: nah we just don't need these

**Comment by RubenMattan** [comment]
migrations:
- 6 in this pull request
- provision all services
- remove all events from news stream layouts

**Pull request merged**